### PR TITLE
support multiple yaml documents

### PR DESCRIPTION
**Pull request by krzyzacy**
alternative of #8313 

/area prow
/assign @BenTheElder @fejta @cjwagner

### CODE REVIEW

File: prow/config/config.go [hunk 1]
```diff
@@ -286,42 +285,26 @@ func Load(prowConfig, jobConfig string) (*Config, error) {
...
+	// split into multiple blobs :-)
+	jobData, err := ioutil.ReadFile(jobConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error reading %s: %v", jobConfig, err)
+	}
 
-		if info.IsDir() {
-			return nil
-		}
+	jobBlobs := strings.Split(string(jobData), "---")
```
Comment (BenTheElder) [inline prow/config/config.go]: this isn't quite right. it needs to split on lines that start with `---` at the beginning of the line, `---` is valid in other places like `- foo: "bar --- baz"`

also once split, we need to gracefully handle:
```yaml
---
- foo: bar
   baz: quux
---
- this: is
   fine: doge
---
```
(IE empty ranges)

Comment (krzyzacy) [reply prow/config/config.go]: > I would probably do this in a method that generically breaks a byte blob into a slice of yaml-spec-allowed byte slices not-yet deserialized blocks to be deserialized by something aware of the types.

be more specific? Like walk through the file and split by `---\n`? Or reimplement the yaml parser 🤦‍♂️

Comment (BenTheElder) [reply prow/config/config.go]: just the part that splits up the documents. walking all _lines)_ and splitting on lines with `---`. (though if we want it to be more reusable, allowing lines with more than 3 `-` to split would be good (not sure if that's standard, but it seems widely used)

Comment (krzyzacy) [reply prow/config/config.go]: ack sure I can do that

**Comment by fejta** [comment]
Why do we need or want this?

**Comment by krzyzacy** [comment]
@fejta we want to support splitted config files, however k8s configmap does not support subdirs, so here I want to pre-process all the data and concatenate them into one yaml file for config agent to use.

**Comment by fejta** [comment]
Why do we need sub-dirs, as opposed to different files in a single dir? What happens when the configmap key includes a / in the name?

**Comment by krzyzacy** [comment]
@fejta 
```
a valid config key must consist of alphanumeric characters, '-', '_' or '.' (e.g. 'key.name', or 'KEY_NAME', or 'key-name', regex used for validation is '[-._a-zA-Z0-9]+')
```

**Comment by fejta** [comment]
> Why do we need sub-dirs, as opposed to different files in a single dir?

Seems like it would be reasonable to ask people to use unique basenames for their files rather than adding buggy/complicated splitting logic? This is what we do for testgrid

**Comment by krzyzacy** [comment]
@fejta like stub out all the `/` with `_` and dump to configmap?

**Comment by fejta** [comment]
Prow should just concern itself with making the config flag accept multiple arguments like `--config=first-half --config=second-half`.

The mechanics of how the deployment gets `first-half` and `second-half` into the pod is irrelevant to the config code (technically there is no requirement to use a configmap, it could be an init container which curls it from github, for example).

**Comment by krzyzacy** [comment]
we don't want to restart the deployment in order to fetch again use init container that's why we used a configmap right? :-) It can probably just be a pd with the entire `jobs/` dir, but that requires more change to updateconfig plugin.

**Comment by fejta** [comment]
There is no requirement that all config files live inside a single configmap.

For example, right now `prow/config.yaml` and `prow/plugins.yaml` live in the config and plugins configmaps, respectively.

https://github.com/kubernetes/test-infra/blob/2600818521605d935f0636f7a02c3842ffca9101/prow/plugins.yaml#L155-L158

If we split apart `prow/config.yaml` into `foo.yaml`, `bar.yaml` and `frobber.yaml` there's no reason we cannot put these into three different config maps.

This approach does not require any changes to the updateconfig plugin -- https://github.com/kubernetes/test-infra/tree/master/prow/plugins/updateconfig

**Comment by krzyzacy** [comment]
whenever we add a new config file we need to update the deployment to append a flag? that sounds not reasonable.

**Comment by fejta** [comment]
I would be interested in a design doc that outlines the various alternatives you have considered and the assumptions you are using to choose the best way forward here.

**Comment by BenTheElder** [comment]
> It can probably just be a pd with the entire jobs/ dir, but that requires more change to updateconfig plugin.

A PD is probably going to be unnecessarily complex to update. ConfigMaps have some very nice properties around atomic updates as well.

> There is no requirement that all config files live inside a single configmap.

Right, but for the n foo-jobs.yaml files having a configmap for each is probably not the greatest workflow, we'd have to reconfigure the updater every time we added a file and we'd need some naming scheme etc...

Alternatively with something like this, the updater scans all `jobs/*.yaml`, concats them into a file and updates the jobs' single configmap. We should be able to find code for this parsing from say, kubectl, but it also shouldn't be very hard to just implement splitting / concatting..

Splitting yaml documents in one stream/file it is actually just splitting on lines with `^---+$` and is in the yaml spec.


Not sure if there are other good options we're missing or if this is the best, but a stream of documents in one file is a pretty standard Kubernetes pattern. 

An uglier option is to autogenerate a mapping of job files to mangled configmap entry names and use a ton of `--from-file=mangle(filename)=filename`, or to require handwriting the mapping every time we split the jobs into a new file.

**Comment by krzyzacy** [comment]
fixed the file split part, will add a design doc and some more unit tests afterwards.

**Comment by fejta** [comment]
> whenever we add a new config file we need to update the deployment to append a flag? that sounds not reasonable.

> n foo-jobs.yaml files having a configmap for each is probably not the greatest workflow

This may or may not be a big deal depending on the size of  n and/or the frequency we rename `foo.yaml` into `bar.yaml` and/or split `foo.yaml` into `foo-x.yaml` and `foo-y.yaml`

A problem with trying to put everything inside a single configmap (regardless of the number of keys used) is that it breaks kubectl apply.

Allowing different configmaps make it easier for different ones to live in different repos.

> An uglier option is to autogenerate a mapping of job files to mangled configmap entry names and use a ton of --from-file=mangle(filename)=filename, or to require handwriting the mapping every time we split the jobs into a new file.

I don't see the need to mangle the filename. Just give an approximate solution: require that basename(filename) is unique for all files and use that as the key.

**Comment by BenTheElder** [comment]
We have to also require that basename is from the (quite limited) valid
character set unless we mangle.

Do we need kubectl apply for job config configmap?

**Comment by krzyzacy** [comment]
same question, why kubectl apply will break?

For contributor's workflow, they should be able to add a new yaml file for their sig/repo and not worry about other prow configs. They will have their OWNERS file under their subdir so that if they only touches their own configs then we don't have to review their changes from the prow side.

File: prow/config/config.go [hunk 2]
```diff
@@ -18,11 +18,11 @@ limitations under the License.
 package config
 
 import (
+	"bufio"
```
Comment (stevekuznetsov) [inline prow/config/config.go]: I read this as `buffalo` and thought we were about to start using the best stdlib library in existence ...

File: prow/config/config.go [hunk 3]
```diff
@@ -286,44 +286,53 @@ func Load(prowConfig, jobConfig string) (*Config, error) {
...
+	file, err := os.Open(yamlPath)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
 
-		subConfig, err := yamlToConfig(path)
-		if err != nil {
-			return err
+	splitRegexp := regexp.MustCompile("^-{3,}$")
```
Comment (stevekuznetsov) [inline prow/config/config.go]: I would be surprised if the YAML unmarshaller doesn't already do this -- the base Golang JSON unmarshaller when taking a stream will give you every JSON document in succession

Comment (krzyzacy) [reply prow/config/config.go]: hummm is it? See https://github.com/kubernetes/test-infra/pull/8313, I saw the streaming decoder is added not long ago, and seems we don't want to bump our yaml library.

**Comment by BenTheElder** [comment]
Bumping out to be fine? Switching libraries might not.

**Comment by BenTheElder** [comment]
Ought*

**Comment by krzyzacy** [comment]
ghodss/yaml is base off go-yaml anyway :-)

**Comment by stevekuznetsov** [comment]
Yeah my only comment here is that if possible let's let the library handle multiple documents in one file, if we need to let's patch our vendor and propose upstream later if we can't bump.

**Comment by fejta** [comment]
We have a lot of testing which just builds parsed structs and skips parsing yaml documents. I think we ought to close those testing holes before we switch out the library.

Two particular things I am concerned about:
* other libraries require you to add yaml annotations rather than just json (this was previously an issue for testgrid), and 
* handle pointer values differently (currently we can distinguish between default value and unset by using a pointer, previously we could not)

**Comment by krzyzacy** [comment]
@fejta I'm using yaml library from k8s apimachinery now, which uses ghodss/yaml in behind and implements the streaming decoder on top. 

https://github.com/kubernetes/apimachinery/blob/master/pkg/util/yaml/decoder.go#L29

**Comment by krzyzacy** [comment]
also I'm not touching testgrid :-)

**Comment by krzyzacy** [comment]
ok we are going with multiple keys in configmap for now, gonna close this. But streaming parser might be useful in the future, this PR is a pretty good reference.

**Pull request closed without merging**